### PR TITLE
*: use dumpling's finish building connection location to leave safe mode

### LINKED ISSUE #907: more actions on Dump unit
## Feature Request

**Is your feature request related to a problem? Please describe:**

1. worried that 5 minutes safe mode is not enough when use `--consistency none` to dump data (for example, dump from Aurora)
2. remove ANSI_QUOTES in load unit

**Describe the feature you'd like:**

1. at Close stage, record binlog position (still learning how to deal with GTID set) so could be treated as exit point of safe mode
2. at Init stage, record GLOBAL variables so could be used in load unit

**Describe alternatives you've considered:**

leave to discuss

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve? 
close #907

### What is changed and how it works?

when dump unit can't assure consistency (for example migrate Aurora, or maybe we could support more scenario to avoid FTWRL), between start and ending of dumping's creating connections, binlog maybe duplicated with dump result, so we should enable safe mode in this period. originally safe mode was turned on for 5 minutes, this PR assure safe mode will keep for enough time.

**maybe we should save exit location to checkpoint table, to handle worker switch before exit safe mode. This will achieve in another PR including updating strategy**

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

Side effects

 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to be included in the release note

### CODE REVIEW

**Comment by lance6716** [comment]
waiting dumpling to record exit binlog location in metadata files

**Comment by lance6716** [comment]
waiting https://github.com/pingcap/dumpling/pull/146

File: pkg/dumpling/utils.go [hunk 1]
```diff
@@ -0,0 +1,181 @@
+// Copyright 2019 PingCAP, Inc.
```
Comment (lance6716) [inline pkg/dumpling/utils.go:1]: renamed from `pkg/utils/mydumper.go` to avoid cycle import

**Comment by lance6716** [comment]
PTAL @csuzhangxc @GMHDBJD

File: syncer/checkpoint.go [hunk 2]
```diff
@@ -805,22 +810,11 @@ func (cp *RemoteCheckPoint) genUpdateSQL(cpSchema, cpTable string, location binl
...
-	if err != nil {
-		return nil, err
-	}
+	loc, dumpExitLoc, err := dumpling.ParseMetaData(filename, cp.cfg.Flavor)
 
-	return &binlog.Location{
-		Position: pos,
-		GTIDSet:  gset,
-	}, nil
+	return loc, dumpExitLoc, err
```
Comment (csuzhangxc) [inline syncer/checkpoint.go:819]: Can `return dumpling.ParseMetaData(filename, cp.cfg.Flavor)` directly.

File: dm/config/task.go [hunk 3]
```diff
@@ -239,6 +240,8 @@ type SyncerConfig struct {
 	EnableGTID       bool `yaml:"enable-gtid" toml:"enable-gtid" json:"enable-gtid"`
 	DisableCausality bool `yaml:"disable-detect" toml:"disable-detect" json:"disable-detect"`
 	SafeMode         bool `yaml:"safe-mode" toml:"safe-mode" json:"safe-mode"`
+	// when dump unit can't run consistent dump, enable safe mode until pass exit location of dumping
+	DumpExitLocation *binlog.Location `yaml:"-" toml:"-" json:"-"`
```
Comment (csuzhangxc) [inline dm/config/task.go:244]: The name of `*Exit*` may let the user think the dump unit is exit/finished (although it is only used internally), but I have no better name. 😹

File: syncer/syncer.go [hunk 4]
```diff
@@ -1094,7 +1094,7 @@ func (s *Syncer) Run(ctx context.Context) (err error) {
 		if s.cfg.Mode == config.ModeAll {
 			if err = s.flushCheckPoints(); err != nil {
 				s.tctx.L().Warn("fail to flush checkpoints when starting task", zap.Error(err))
-			} else {
+			} else if s.cfg.CleanDumpFile {
```
Comment (csuzhangxc) [inline syncer/syncer.go:1097]: this may fix the error in #941?

**Review by csuzhangxc — COMMENTED** [review]
> maybe we should save exit location to checkpoint table, to handle worker switch before exit safe mode.

the case:

1. start task with --consistency none and enter sync stage
2. flush checkpoint for sync unit
3. pause-task
4. resume-task

Should save exit location to checkpoint table?

File: pkg/dumpling/utils.go [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+
+		loc  *binlog.Location
+		loc2 *binlog.Location
+	)
+
+	br := bufio.NewReader(fd)
+
+	parsePosAndGTID := func(pos *mysql.Position, gtid *string) error {
+		for {
+			line, err := br.ReadString('\n')
```
[the remark this answers is not in the document]
Comment (lance6716) [reply pkg/dumpling/utils.go]: https://github.com/pingcap/dm/pull/915/commits/3aa0d9b77babdb5f7be9404a47f1179573e592fb

File: dm/config/task.go (hunk 3 above)
Comment (lance6716) [reply dm/config/task.go]: https://github.com/pingcap/dm/pull/915/commits/16270b591ca3ff2cb491a68759203d19b24a1d97

File: syncer/syncer.go (hunk 4 above)
Comment (lance6716) [reply syncer/syncer.go:1097]: we default enabled `CleanDumpFile`, so if this line was added #941 still happens. Maybe we should check downstream checkpoint, if not exist then load from file 😢

File: syncer/checkpoint.go (hunk 2 above)
Comment (lance6716) [reply syncer/checkpoint.go]: https://github.com/pingcap/dm/pull/915/commits/16270b591ca3ff2cb491a68759203d19b24a1d97

**Comment by lance6716** [comment]
> Should save exit location to checkpoint table?

Yes, going to change checkpoint structure, implement DM version update changes, and https://github.com/pingcap/dm/pull/915#discussion_r481804379 in new PR

File: syncer/syncer.go (hunk 4 above)
Comment (csuzhangxc) [reply syncer/syncer.go:1097]: previous `if fresh` can achieve this?

**Review by csuzhangxc — COMMENTED** [review]
LGTM

**Review by GMHDBJD — APPROVED** [review]
LGTM

**Pull request merged**